July meeting notes (#337)
tc39/tc39-notes: meetings/2024-07/july-29.md

RPR: there is no one in the queue. So if there is any questions, and I will say that thank you for advertising this. Last time we had on the reflector and not everyone saw this, and thank you for this and you can all attend if you wish. On the Tokyo trip. All right we are doing well on the agenda. And onward we go to updates from the code of conduct committee, Chris?

## Updates from the CoC Committee

Presenter: Chris de Almeida (CDA)

- no proposal
- no slides

> CDA: The conduct goes well for the most part. We have not had any new reports and nothing notable in our spaces and disasters course or GetHub.

Reviewer: ```suggestion
CDA: The conduct goes well for the most part. We have not had any new reports and nothing notable in our spaces and discourse or GitHub.
```

a disasters course would be neat tho

> CDA: We have not – we prefer not to be busy on the Code of Conduct committee, as always, I will repeat my pitch for folks who would like to join the Code of Conduct committee and please reach out to one of on the Code of Conduct committee and the main and is that you are available for period of one hour every two weeks. Often we do not have to meet during that hour because we have nothing to talk about but it is important that it is blocked out on your calendar to we can function as a committee when we need to. So thank you.

Reviewer: ```suggestion
CDA: We have not – we prefer not to be busy on the Code of Conduct committee, as always, I will repeat my pitch for folks who would like to join the Code of Conduct committee and please reach out to one of us on the Code of Conduct committee. The main ask is that you are available for a period of one hour every two weeks. Often we do not have to meet during that hour because we have nothing to talk about but it is important that it is blocked out on your calendar so we can function as a committee when we need to. So thank you.
```

USA: Thank you, Shu. So I think we could capture the queue from today and come back to this.

> DE: so as part of the conclusion say people disagree with this as implementation defined engage before the next meeting? This is a request from SYG and this included WH and maybe other people on the queue and MLS.

Reviewer 2: ```suggestion
DE: Can we agree, as part of the conclusion, that people who disagree with calling the behavior "implementation-defined" will engage with SYG offline before the next meeting? e.g., WH, MLS.
```

MF: Um, no that is the number of –

JHD: Oh okay. Got it. Okay, so the concurrency and methods, the methods below that, are not directly related to the buffered helper, that was my confusion?

MF: No, no that concurrency parameter is how those methods are driving the underlying iterator and so again how many outstanding promises they will keep.

JHD: Okay. Thank you.

USA: And we are on time? Okay let’s see. There is nothing in the queue.

MF: Okay great! If you think of anything else, please open an issue. Thank you.

USA: Would you like to dictate a summary?

MF: Not really?

USA: Could you edit it into the notes?

MF: Sure. Sure I will write it in the notes.

### Speaker's Summary of Key Points

- concurrency control advanced to Stage 1
- there was much concern about the naming of Semaphore
- there was some skepticism about the motivation for the full Governor abstraction from Igalia and Mozilla
- MM had concerns about cross-Agent sharing of Semaphores

## AsyncContext Presenter: Chengzhong Wu, Andreu Botella

CZW: All right, I will share my screen. It is black for me, I don’t know.

> ABO: This is Andreu, and so this the AsyncContext update about integration, and some – next slide.DE: If I can jump in, I got in touch with some framework feedback about AsyncContext, and people are generally very excited about this feature, and both for client and server and so discussed some Dev’s, and keeping track of what the component is as well as performance tracing and I had some hopes that AsyncContext would make sense for abrupt context file variables but these are already handled very well by referencing the tree, there is a lot of demand for that in particular. But I think it is safe to say this feature would be to positive effects for allowing adoption of async await frameworks.

Reviewer 2: ```suggestion
ABO: This is Andreu, and so this the AsyncContext update about integration, and some – next slide.

DE: If I can jump in, I got in touch with some framework maintainers about AsyncContext, and people are generally very excited about this feature, and both for client and server. Some use cases include
- keeping track of what the component is
- performance tracing
- maintaining hooks state
- Generally, allowing adoption of async/await in frameworks

I had some hopes that AsyncContext would make sense for the equivalent of React Context, but these are already implemented efficiently by referencing the component tree, so there is not a lot of demand for that particular usage mode.
```

ABO: So, um, the like whenever you have like promise you set up a promise resolution that like – the only thing that happens when you have like when you have like a promise resolution, this already happens like the spec calls into a host hook to like queue the rejection event. And this would like 100%. And so creating like new context at that point and expose it on the rejection event. And for hosts that does not have reaction events and that simple exit the process about there is a rejection this would not have any change in behaviour or any global state.

MM: So with an unhandled rejection, the host action of unhandled rejection that sounds reasonable but the unhandled rejection does not happen at the point of rejection. It happens at the point where the language notices that the rejection was not handled. You need to capture the context where the rejection happened, right?

DE: Can I address here your question Mark, and I am on the queue?

MM: Please.

> DE: So, yes, for promises, promise end is consistently handled as well as promise catch, there consistently handles with in the context snapshot where the reaction was scheduled. So whether you do a weight or you do then or catch or finally, it is always withdrawing that same one. Now our behaviour with callback in the web platform is to basically do that as well. Just as easily as promises, they restore that snapshot. Sometimes there is other contextual information that is relevant. And so, the solution we are going for and rather than doing something fancy and automatic is to make it so that it is as if you constructed a new AsyncContext snapshot at the point when some interesting thing happened. So in this case, it is for – when you reject a promise, the snapshot is sort of eagerly captured. And then if it turns out it is not handled, then um, then that will get passed to the event. An alternative semantics when promise is allocated and capture snapshot in case it is an unhandled rejection. These two sound different but turns out that they are actually similar. And we were not really able to find that cases that one behaved differently than the other for developers and we are proposing where it is at the time of rejection. Though for example the system then, a promise you know and there is no JavaScript when it is rejected, and then it has to have been captured earlier. For example when the promise is allocated.

Reviewer 2: ```suggestion
DE: So, yes, for promises, `Promise.then` is consistently handled as well as `Promise.catch`: both reactions consistently run within the context snapshot where the reaction was scheduled. So whether you do `await` or you do `.then` or `.catch` or `.finally`, it is always executing within that "registration" context.

DE: Now, our proposed behaviour with callbacks in the web platform is to basically do that as well. Just like Promises, they restore the context snapshot which was active when the callback was received as an argument. Sometimes there is other contextual information that is relevant. And so, the solution we are going for and rather than doing something fancy and automatic is to make it so that it is as if you constructed a new AsyncContext.Snapshot at the point when some interesting thing happened, and that is made available somewhere else for explicit use.

DE: For example, when you reject a promise, the snapshot is sort of eagerly captured. And then if it turns out it is not handled, then that snapshot will get passed to the event handler, exposed as a property on the Event object.

DE: An alternative semantic is, when promise is allocated and capture snapshot in case it is an unhandled rejection. These two sound different but turns out that they are actually similar. And we were not really able to find that cases that one behaved differently than the other for developers and we are proposing where it is at the time of rejection. Note that, when the platform has an API which returns a Promise, there might be no JavaScript active when the Promise is rejected, and so the platform must cache the Snapshot earlier, e.g., when the Promise is allocated.
```

MM: You should note by the way the idea of capturing these is attractive to me. I like the idea of it. I am just – what I am raising is how it fits with the rest of the SA men ticks and you answered that very well, and I will clarify two things. And that it is only – if you are capturing the rejection, it is only want rejection, and not on fulfillment.

> DE: That is right. The argument is that we do this on fulfillment as well and that is not currently being proposed, what is being proposed is unfulfillment and nothing is captured.

Reviewer 2: ```suggestion
DE: That is right. Our proposal is that, on fulfillment, nothing is captured.
```

KKL: I want to – I am on the queue. In the previous life I was involved in the open context project, and precursor to open tel., and so this comes to propagation that will come to bear on this, and this is interesting because reverse context propagation reverses the arrow in the way that goes from one to many to many to one. Which sometimes implies that you have to use completely different data structures to pull data out of reverse context. I don’t know whether this comes to bear on this, and really just wanted to propose that we talk through it on TG3, and if you can add yourself to a future agenda that would be great. The thing about reverse context that happens if you fanout to multiple parties and any of them might return reverse context is that the return context has to be aggregated some way in order to be made sense of at the conception site. So get – the get on the last line of this example would effectively require a reduction of possibly many changes. And in any case, again I don’t know whether that comes to bear, but it would be great to talk through.

> DE: Yeah, we have had a bit of debate about this within the group, and I agree that there are some multiple context that are relevant. Which is kind of why we would hoping to go with this simple policy. But yeah, let’s get on the TG3 ideas.

Reviewer 2: ```suggestion
DE: Yeah, we have had a bit of debate about this within the group, and I agree that there are some multiple context that are relevant. Which is kind of why we would hoping to go with this simple policy. But yeah, let’s get on the TG3 agenda.
```
